scrapy/utils/gz.py

```diff
@@ -4,30 +4,38 @@
     from cStringIO import StringIO as BytesIO
 except ImportError:
     from io import BytesIO
-
+from io import UnsupportedOperation
 from gzip import GzipFile
 
+class ReadOneGzipFile(GzipFile):
+    def readone(self, size=-1):
+        try:
+            return self.read1(size)
+        except UnsupportedOperation:
+            return self.read(size)
```

Comment:
@kmike ,
yes, in Python 3.5, with current code and corrupted input, one gets `EOFError` but missing data.

For `GunzipTest.test_gunzip_truncated`, the first `read(8196)` call gets first chunk, but second call raises `EOFError('Compressed file ended before the end-of-stream marker was reached',)` and missing data fails the assertion on test content.

For `GunzipTest.test_gunzip_truncated_short`, `read(8196)` raises `EOFError` for first chunk.

I believe this is the [change in Python 3.5](https://docs.python.org/3.5/whatsnew/changelog.html):
`Issue #1159051: GzipFile now raises EOFError when reading a corrupted file with truncated header or footer.`

The only way I found to make it read uncompressed bytes for corrupted input was to use `read1()` -- which is only available from Python 3.3 on `GzipFile`, so that even if `size` bytes are not read, it would still return them before `EOFError` on the next call.

With `read1(8196)`, you get successful calls with chunks of data before raising `EOFError` on last call:

```
GunzipTest.test_gunzip_truncated:
read1(8196) --> 8196
read1(8196) --> 954
read1(8196) --> EOFError
```

```
GunzipTest.test_gunzip_truncated_short:
read1(8196) --> 3741
read1(8196) --> EOFError
```

This is how `read()` mainly differs IMO ([from Python 3 io docs](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read)):

> If the argument is positive, and the underlying raw stream is not interactive, multiple raw reads may be issued to satisfy the byte count (**unless EOF is reached first**)

(Although I'm not that happy with this `readone` definition trying if `read1` is available ([it appeared in Python 3.3](https://docs.python.org/3/library/gzip.html#gzip.GzipFile)))